src/Chapter3.hs

```diff
@@ -479,6 +523,36 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+type People = Int
+
+data Castle = Castle String | NoCastle deriving (Show)
+data Wall = Wall | NoWall deriving (Show)
+data SpecBuilding = Church | Library deriving (Show)
+data House = House People deriving (Show)
+
+data City = City
+  { cityCastle          :: Castle
+  , cityWall            :: Wall
+  , citySpecialBuilding :: SpecBuilding
+  , cityHouses          :: [House]
+  } deriving (Show)
+
+buildCastle :: City -> Castle -> City
+buildCastle city castle = city { cityCastle = castle }
+
+buildHouse :: City -> House -> City
+buildHouse city house = city { cityHouses = house:houses }
+  where houses = cityHouses city
+
+buildWalls :: City -> Wall -> City
+buildWalls city@(City NoCastle _ _ _) _ = city
+buildWalls city wall
+  | isEnoughPeople = city { cityWall = wall }
+  | otherwise  = city
+  where
+    isEnoughPeople = totalPeopleCount >= 10
+    totalPeopleCount = foldl (\acc (House p) -> acc + p) 0 houses
```

Comment:
You can use `sum` in here 🙂 
`foldl` is very lazy function and very slow. You should usually avoid it. You can use `foldl'` instead 👍🏼